core/src/main/java/io/grpc/internal/CallCredentialsApplyingTransportFactory.java

```diff
@@ -147,6 +157,9 @@ public Attributes getTransportAttrs() {
     @Override
     public void shutdown(Status status) {
       checkNotNull(status, "status");
+      if (shutdownInvoked.compareAndSet(false, true)) {
+        pendingApplier.addAndGet(Integer.MAX_VALUE);
```

Comment:
Move this into the synchronized block and change the `if` to check that it is negative. At that point you can delete `shutdownInvoked`.